_drafts/learn-doobie-for-great-good.md

```diff
@@ -0,0 +1,961 @@
+
+```scala
+val name = "John Travolta"
+
+// This statement...
+sql"insert into actors (name) values ($name)".update.run
+
+// ...is equivalent to this one:
+val stmt = "insert into actors (name) values (?)"
+Update[String](stmt).run(name)
+```
+
+Now, we can use the desugared version of the `sql` interpolator to insert multiple rows. As an example, we can insert a list of actors into the database:
+
+```scala
+def saveActorsProgram(actors: NonEmptyList[String]): IO[Int] = {
+  val insertStmt: String = "insert into actors (name) values (?)"
+  val numberOfRows: doobie.ConnectionIO[Int] = Update[String](insertStmt).updateMany(actors.toList)
+  numberOfRows.transact(xa)
+}
+```
+
+As we can see, Doobie gives us a method, `updateMany`, to execute a batch insertion. It takes a list of parameters and returns the number of rows inserted. Some databases, such as Postgres, can produce a list of column values of the inserted rows. In this case, we can use the method `updateManyWithGeneratedKeys`, taking as input a list of column names:
+
+```scala
+def saveActorsAndReturnThemProgram(actors: NonEmptyList[String]): IO[List[Actor]] = {
+  val insertStmt: String = "insert into actors (name) values (?)"
+  val actorsIds = Update[String](insertStmt).updateManyWithGeneratedKeys[Actor]("id", "name")(actors.toList)
+  actorsIds.compile.toList.transact(xa)
+}
+```
+
+Updating information in the database is the same affair as inserting. There are no substantial differences between the two. Imagine we want to fix the year of production of the movie "Zack Snyder's Justice League". Here is how we can do it:
+
+```scala
+def updateJLYearOfProductionProgram(): IO[Int] = {
+  val year = 2021
+  val id = "5e5a39bb-a497-4432-93e8-7322f16ac0b2"
+  sql"update movies set year_of_production = $year where id = $id".update.run.transact(xa)
+}
+```
+
+Finally, the deletion follows the same pattern. Use the keyword `delete` instead of `insert` or `update` inside the `sql` interpolator.
+
+## 6. Doobie's Type Classes
+
+So far, we have seen many examples of usages of the `sql` interpolator, which magically can convert Scala types into JDBC types when reading input parameters, and vice versa when concerning mapping values extracted from the database.
+
+As we can imagine, there is no magic whatsoever. As skilled Scala developers, we should have known that **there are some type classes behind it whenever someone talks about magic**.
+
+In fact, Doobie basically uses four type classes for the conversion between Scala and JDBC types: `Get[A]`, `Put[A]`, `Read[A]` and `Write[A]`.
+
+The `Get[A]` describes creating the Scala type `A` from a non-nullable schema type. We can also apply the same type class to obtain type `Option[A]`. So, Doobie uses an instance of `Get[A]` in mapping the results of a query into Scala.
+
+The `Put[A]` type class describes translating a Scala type `A` into a non-nullable schema type. As for the `Get[A]` type class, the `Put[Option[A]]` comes freely.
```

Comment:
This one isn't quite clear. What do you mean by translating A into a non-nullable schema type?
Also, what do you mean by Put[Option[A]] comes freely?